src/samples/imageCaching/main_imageCaching.cpp

```diff
@@ -0,0 +1,103 @@
+// This file is part of the AliceVision project.
+// Copyright (c) 2022 AliceVision contributors.
+// This Source Code Form is subject to the terms of the Mozilla Public License,
+// v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include <aliceVision/image/caching.hpp>
+
+#include <aliceVision/system/Logger.hpp>
+#include <aliceVision/system/cmdline.hpp>
+#include <aliceVision/system/main.hpp>
+
+#include <boost/program_options.hpp>
+
+#include <thread>
+#include <mutex>
+#include <functional>
+#include <chrono>
+
+
+// These constants define the current software version.
+// They must be updated when the command line is changed.
+#define ALICEVISION_SOFTWARE_VERSION_MAJOR 1
+#define ALICEVISION_SOFTWARE_VERSION_MINOR 0
+
+using namespace aliceVision;
+using namespace std::chrono_literals;
+
+namespace po = boost::program_options;
+
+
+int aliceVision_main(int argc, char **argv)
+{
+    // command-line arguments
+    int capacity = 256;
+    int maxSize = 1024;
+    std::vector<std::string> filenames;
+    std::vector<int> halfSampleLevels;
+    int nbThreads = 1;
+
+    po::options_description requiredParams("Required parameters");
+    requiredParams.add_options()
+        ("filenames", po::value<std::vector<std::string>>(&filenames)->multitoken()->default_value(filenames), 
+        "Filenames")
+        ;
+
+    po::options_description optionalParams("Optional parameters");
+    optionalParams.add_options()
+        ("capacity", po::value<int>(&capacity)->default_value(capacity), 
+        "Cache capacity")
+        ("maxSize", po::value<int>(&maxSize)->default_value(maxSize), 
+        "Cache max size")
+        ("halfSampleLevels", po::value<std::vector<int>>(&halfSampleLevels)->multitoken()->default_value(halfSampleLevels), 
+        "Half-sampling levels")
+        ("nbThreads", po::value<int>(&nbThreads)->default_value(nbThreads), 
+        "Number of threads")
+        ;
+
+    CmdLine cmdline("AliceVision imageCaching");
+    cmdline.add(requiredParams);
+    cmdline.add(optionalParams);
+    if (!cmdline.execute(argc, argv))
+    {
+        return EXIT_FAILURE;
+    }
+
+    // Setup cache
+    ALICEVISION_LOG_INFO("Creating image cache with capacity " << capacity << " and max size " << maxSize);
+    image::ImageReadOptions readOptions(image::EImageColorSpace::LINEAR);
+    image::ImageCache cache(capacity, maxSize, readOptions);
+
+    // Start threads
+    std::vector<std::thread> threads;
+    std::mutex mutexLoadImg;
+    for (int i = 0; i < nbThreads; i++)
+    {
+        threads.emplace_back([&](int numThread){
+            // Load images
+            for (int j = numThread; j < filenames.size(); j += nbThreads)
+            {
+                {
+                    const std::lock_guard<std::mutex> lock(mutexLoadImg);
```

Comment:
This lock ensures that regardless of the number of threads, exactly one thread will be accessing the image cache at a time.

Reply:
Ah right my bad, I'll just remove it